libsolidity/ast/AST.cpp

```diff
@@ -243,6 +243,41 @@ vector<EventDefinition const*> const ContractDefinition::usedInterfaceEvents() c
 	);
 }
 
+vector<EventDefinition const*> ContractDefinition::interfaceEvents(bool _requireCallGraph) const
+{
+	set<EventDefinition const*, CompareByID> result;
+	for (ContractDefinition const* contract: annotation().linearizedBaseContracts)
+		result += contract->events();
+	solAssert(annotation().creationCallGraph.set() == annotation().deployedCallGraph.set());
+	if (_requireCallGraph)
+		solAssert(annotation().creationCallGraph.set());
+	if (annotation().creationCallGraph.set())
+		result += usedInterfaceEvents();
+	// We could filter out all events that do not have an external interface
+	// if _requireCallGraph is false.
+	return util::convertContainer<vector<EventDefinition const*>>(std::move(result));
+}
+
+vector<EventDefinition const*> ContractDefinition::natspecInterfaceEvents() const
+{
+	auto eventSignature =
+		[](EventDefinition const* _event) -> string {
+			FunctionType const* functionType = _event->functionType(true);
+			solAssert(functionType, "");
+			return functionType->externalSignature();
+	};
+	auto compareBySignature =
+		[&eventSignature](EventDefinition const* _lhs, EventDefinition const* _rhs) -> bool {
+			return eventSignature(_lhs) < eventSignature(_rhs);
+	};
+
+	set<EventDefinition const*, decltype(compareBySignature)> events{compareBySignature};
+	events += definedInterfaceEvents();
+	events += usedInterfaceEvents();
```

Comment:
I think it would be a good idea to comment that the order here is very important:

```suggestion
	set<EventDefinition const*, decltype(compareBySignature)> events{compareBySignature};
	// NOTE: Insert events defined in the contract first so that in case of a conflict
	// they're the ones that get selected.
	events += definedInterfaceEvents();
	events += usedInterfaceEvents();
```